lib/engine/game/g_1828.rb

```diff
@@ -221,6 +230,131 @@ def upgrades_to?(from, to, special = false)
+        return [] if !player || !corporation
+        return [] if corporation.system?
+
+        @corporations.select do |candidate|
+          next if candidate == corporation ||
+                  candidate.system? ||
+                  !candidate.ipoed ||
+                  (corporation.owner != player && candidate.owner != player) ||
+                  candidate.operated? != corporation.operated? ||
+                  (!candidate.floated? && !corporation.floated?)
+
+          # account for another player having 5+ shares
+          @players.any? do |p|
+            num_shares = p.num_shares_of(candidate) + p.num_shares_of(corporation)
+            num_shares >= 6 ||
+              (num_shares == 5 && !sold_this_round?(p, candidate) && !sold_this_round?(p, corporation))
+          end
+        end
+      end
+
+      def sold_this_round?(entity, corporation)
+        return false unless @round.players_sold
+
+        @round.players_sold[entity][corporation]
+      end
+
+      def create_system(corporations)
+        system_data = CORPORATIONS.find { |c| c['sym'] == corporations.first.id }.dup
+        system_data['sym'] = corporations.map(&:name).join('-')
+        system_data['tokens'] = []
+        system = init_corporation(@stock_market, system_data)
+        system.extend(Engine::G1828::System)
+        system.name = corporations.first.name
+
+        @corporations << system
+        @_corporations[system.id] = system
+        system.shares.each { |share| @_shares[share.id] = share }
+
+        create_system_tokens(system, corporations)
+
+        # Make sure the system will not own two coal markers
+        if corporations.count { |corp| coal_marker?(corp) } == 2
+          remove_coal_marker(corporations.last)
+          add_coal_marker_to_va_coalfields
+          @log << "#{system.name} cannot have two coal markers, returning one to Virginia Coalfields"
+        end
+
+        corporations.each do |corporation|
+          system.corporations << corporation
+          corporation.spend(corporation.cash, system) if corporation.cash.positive?
+
+          shell = Engine::G1828::Shell.new(corporation.name, system)
+          corporation.trains.dup.each do |train|
+            system.buy_train(train, :free)
+            shell.trains << train
+          end
+          system.shells << shell
+
+          corporation.companies.each do |company|
+            company.owner = system
+            system.companies << company
+          end
+          corporation.companies.clear
+
+          corporation.all_abilities.each do |ability|
+            system.add_ability(ability)
+            @target.remove_ability(ability)
+          end
+        end
+
+        @stock_market.set_par(system, system_market_price(corporations))
+        system.ipoed = true
+
+        system
+      end
+
+      def create_system_tokens(system, corporations)
+        used, unused = corporations.map(&:tokens).flatten.partition(&:used)
```

Comment:
flat_map

Reply:
One day I'll remember that method exists. ;-)